packages/astro/src/core/build/generate.ts

```diff
@@ -39,26 +39,41 @@ import { getTimeStat } from './util.js';
 // system, possibly one that parallelizes if async IO is detected.
 const MAX_CONCURRENT_RENDERS = 1;
 
-// Throttle the rendering a paths to prevents creating too many Promises on the microtask queue.
-function* throttle(max: number, inPaths: string[]) {
-	let tmp = [];
-	let i = 0;
-	for (let path of inPaths) {
-		tmp.push(path);
-		if (i === max) {
-			yield tmp;
-			// Empties the array, to avoid allocating a new one.
-			tmp.length = 0;
-			i = 0;
-		} else {
-			i++;
-		}
+async function* generatePathsConcurrently(
+	concurrency: number,
+	opts: StaticBuildOptions,
+	gopts: GeneratePathOptions,
+	timeStart: number,
+	pathsIterator: Iterable<string>,
+	routeType: RouteType
+) {
+	// Each async iterator is a worker.
+	// Sharing the iterator ensures that each worker gets unique tasks.
+	const asyncIterators = new Array<AsyncGenerator<void, void, unknown>>(concurrency);
+	for (let i = 0; i < concurrency; i++) {
```

Comment:
I'm not fully understanding the logic here. If concurrency is say 5 and there are 10 paths, how/when does it generate paths 5-10? It appears to me that it stops and only does the first 5. What am I missing?

Reply:
No, if concurrency is 5 it creates 5 AsyncGenerator, each of them shares the `pathsIterator` which contains all the paths we generate (in my case that's in the thousands).

So basically you have 5 "workers" and a "shared queue", each "worker" calls next() on the pathsIterator and processes that path until all the paths are processed and the pathsIterator is exhausted.

Reply:
Ah ok, so I think I understand this now.